sql: improve SHOW TABLES to show row count

ekalinin opened the pull request:
Fixes #48755

This change adds column `estimated_row_count` into result of the "SHOW TABLES"
to show estimated (not real) number of rows.

Besides that the body of the `delegator.delegateShowTables` was a bit
simplified (we do not need two almost similar versions of the query) and
new virtual table `crdb_internal.table_row_statistics` was added (to
show stats for non-root users).

Release note (sql change): This change modifies `SHOW TABLES` to return
estimates number of rows. New column's name is `estimated_row_count`.
Number of rows is taken from `system.table_statistics` table (via
`crdb_internal.table_row_statistics` which shows only tables accessible
for current user).

ekalinin committed:
sql: improve SHOW TABLES to show row count

This change adds column "estimated_row_count" into result of the "SHOW TABLES"
to show estimated (not real) number of rows.

Besides that the body of the `delegator.delegateShowTables` was a bit
simplified (we do not need two almost similar versions of the query) and
new virtual table `crdb_internal.table_row_statistics` was added (to
show stats for non-root users).

Release note (sql change): This change modifies SHOW TABLES to return
estimates number of rows. New column's name is `estimated_row_count`.
Number of rows is taken from `system.table_statistics` table (via
`crdb_internal.table_row_statistics` which shows only tables accessible
for current user).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52203)
<!-- Reviewable:end -->

rohany commented:
Thanks for the contribution @ekalinin! I gave this a try earlier and ran into a problem that I think your implementation here also hits. The problem is that only the admin user is allowed to access the `system.table_statistics` table. A non-admin user is going to get a permission denied error running `SHOW TABLES` even if they have permissions to see all of the tables that they want to access statistics on. It is a limitation with the current way we do these delegate statements. cc @jordanlewis this is what I was talking about re delegate permissions are annoying.

rohany commented:
A potential solution is to implement a new crdb internal virtual table that exposes the latest statistics on tables, and then restricts its own output to only the tables that the current user has privileges to access. Finally, the delegate command can join against it.

If this sounds like something you want to take on, I can link a commit that adds a new virtual table.

knz commented on pkg/sql/delegate/show_tables.go:
```diff
@@ -43,53 +43,35 @@ func (d *delegator) delegateShowTables(n *tree.ShowTables) (tree.Statement, erro
+       END                                AS "type"
+     , (select max("rowCount")
+          from system.table_statistics
+         where "tableID" = pc.oid::int)   AS "rows"
```
When this PR or something similar goes in, please make sure that the column name properly reflects that this an estimate. This current name is misleading - we don't want users coming to our tech support asking why the value is "wrong".

ekalinin replied:
Yeah, good point. Thanks.
What about "est.rows"?

knz replied:
`row_count_estimate` sounds like a better plan.

ekalinin replied:
Fixed.

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52203)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @ekalinin, and @rohany)

___
*[pkg/sql/delegate/show_tables.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52203#-MDkSBw540f0JEmgdgTJ:-MDkSBw540f0JEmgdgTK:b-6sqrpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/36c0f80ce2ad214ed39741cc6ec4a81f1555b944/pkg/sql/delegate/show_tables.go#L54)):*
> ```Go
>             ELSE 'table'
>        END                                AS "type"
>      , (select max("rowCount")
> ```

The maximum may not be the most accurate. You'll probably want to look at the "createdAt" column to get the most recent value for the given table.


<!-- Sent from Reviewable.io -->

ekalinin commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52203)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @ekalinin, and @rohany)
> 
> _[pkg/sql/delegate/show_tables.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52203#-MDkSBw540f0JEmgdgTJ:-MDkSBw540f0JEmgdgTK:b-6sqrpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/36c0f80ce2ad214ed39741cc6ec4a81f1555b944/pkg/sql/delegate/show_tables.go#L54)):_
> 
> > ```go
> >             ELSE 'table'
> >        END                                AS "type"
> >      , (select max("rowCount")
> > ```
> 
> The maximum may not be the most accurate. You'll probably want to look at the "createdAt" column to get the most recent value for the given table.

Thanks. Fixed.

ekalinin commented:
> A potential solution is to implement a new crdb internal virtual table that exposes the latest statistics on tables, and then restricts its own output to only the tables that the current user has privileges to access. Finally, the delegate command can join against it.
> 

@rohany sounds like a plan :).

> If this sounds like something you want to take on, I can link a commit that adds a new virtual table.

Looks good to me. Could you link that commit, please?

Thanks.

ekalinin commented:
> Thanks for the contribution @ekalinin! I gave this a try earlier and ran into a problem that I think your implementation here also hits. The problem is that only the admin user is allowed to access the `system.table_statistics` table. A non-admin user is going to get a permission denied error running `SHOW TABLES` even if they have permissions to see all of the tables that they want to access statistics on. It is a limitation with the current way we do these delegate statements.

@rohany  Are there any contraindications for allowing all users to read `system.table_statistics`?

I've updated the tests with this logic (moved a table into a separate db and added `GRANT SELECT ON system.table_statistics`):

```
statement ok
CREATE DATABASE rowtest

statement ok
GRANT CREATE, INSERT, SELECT, DELETE ON DATABASE rowtest TO testuser

statement ok
GRANT SELECT ON system.table_statistics TO testuser

user testuser

statement ok
CREATE TABLE rowtest.t1 (a INT)

statement ok
INSERT INTO rowtest.t1 SELECT a FROM generate_series(1, 1024) AS a(a)

query T
SELECT rows FROM [SHOW TABLES from rowtest]
----
NULL

statement ok
ANALYZE rowtest.t1

query I
SELECT rows FROM [SHOW TABLES from rowtest] where table_name = 't1'
----
1024

...
``` 

And it works good.

(see full test here: https://github.com/cockroachdb/cockroach/pull/52203/files#diff-cdf15f008dd1eaef8d7900a2736e8a97 )

rohany commented:
> Looks good to me. Could you link that commit, please?

https://github.com/cockroachdb/cockroach/pull/49771

> Are there any contraindications for allowing all users to read system.table_statistics?

Yes -- we can't allow users to get information about tables that they don't have permissions for.

ekalinin commented:
> > Looks good to me. Could you link that commit, please?
> 
> #49771
>

Thanks. 

Just added a `crdbInternalTablesTableLastStats (crdb_internal.table_row_stats)`. PTAL.
Am I heading in the right direction?
(tests are not fixed yet)
 
> > Are there any contraindications for allowing all users to read system.table_statistics?
> 
> Yes -- we can't allow users to get information about tables that they don't have permissions for.

Got it.

rohany commented on pkg/sql/crdb_internal.go:
```diff
@@ -337,6 +338,57 @@ CREATE TABLE crdb_internal.tables (
+  table_name       STRING      NOT NULL,
+  row_count        INT         NOT NULL
+)`,
+	generator: func(ctx context.Context, p *planner, dbContext *sqlbase.ImmutableDatabaseDescriptor) (virtualTableGenerator, cleanupFunc, error) {
```
you can just implement `populate` here, the API is a bit simpler.

ekalinin replied:
Fixed.

rohany commented on pkg/sql/crdb_internal.go:
```diff
@@ -337,6 +338,57 @@ CREATE TABLE crdb_internal.tables (
+var crdbInternalTablesTableLastStats = virtualSchemaTable{
+	comment: "the latest stats for all tables accessible by current user in current database (KV scan)",
+	schema: `
+CREATE TABLE crdb_internal.table_row_stats (
```
nit -- i'd name this `table_row_statistics`

ekalinin replied:
Renamed.

rohany commented on pkg/sql/crdb_internal.go:
```diff
@@ -337,6 +338,57 @@ CREATE TABLE crdb_internal.tables (
+CREATE TABLE crdb_internal.table_row_stats (
+  table_id         INT         NOT NULL,
+  table_name       STRING      NOT NULL,
+  row_count        INT         NOT NULL
```
I'd name this `estimated_row_count` or something like that, just so it's clear that it is an estimate.

ekalinin replied:
Renamed.

rohany commented on pkg/sql/crdb_internal.go:
```diff
@@ -337,6 +338,57 @@ CREATE TABLE crdb_internal.tables (
+				func(db *sqlbase.ImmutableDatabaseDescriptor, _ string, table *ImmutableTableDescriptor) error {
+					tableID := tree.DInt(table.ID)
+
+					query := `SELECT max("rowCount")` +
```
hmm, the issue with doing it like this is that we have to make a query for each row, which is really unfortunate. I think that you could try to collect all of the latest stats up front into a map and then look up into that during this iteration. However, that has a problem with memory -- if the set of tables is too big then we'll OOM accessing this map... I think that would be unlikely however since all we'd be requesting are 3 `int`s per table.

ekalinin replied:
Yeah, good point. Fixed.

rohany reviewed (changes requested):
Yup, this is definitely on the right track!

ekalinin commented:
Fixed some tests. PTAL.

rohany commented on pkg/sql/crdb_internal.go:
```diff
@@ -337,6 +338,59 @@ CREATE TABLE crdb_internal.tables (
+	populate: func(ctx context.Context, p *planner, db *sqlbase.ImmutableDatabaseDescriptor, addRow func(...tree.Datum) error) error {
+		// Collect the latests statistics for all tables.
+		query := `` +
+			`SELECT s."tableID", max(s."rowCount")` +
```
nit you can just put this query in a 
```
`
multiline
string
`
```

and have it formatted via `sqlfum.pt`

ekalinin replied:
Done (used multiline string, query was re-formatted).

rohany commented on pkg/sql/delegate/show_tables.go:
```diff
@@ -43,53 +43,37 @@ func (d *delegator) delegateShowTables(n *tree.ShowTables) (tree.Statement, erro
-WHERE pc.relkind IN ('r', 'v', 'S') %[2]s
-ORDER BY schema_name, table_name
+	const getTablesQuery = `
+SELECT ns.nspname                         AS schema_name
```
Have this query formatted with sqlfum.pt as well

ekalinin replied:
Done.

rohany commented on pkg/sql/delegate/show_tables.go:
```diff
@@ -43,53 +43,37 @@ func (d *delegator) delegateShowTables(n *tree.ShowTables) (tree.Statement, erro
+            WHEN pc.relkind = 'S' THEN 'sequence'
+            ELSE 'table'
+       END                                AS type
+     , (
```
can you double check that the plan here seems reasonable (i.e. we can decorrelate this subquery into a join)

ekalinin replied:
Fixed. Now it's a left outer join (not a subquery).

rohany commented on pkg/sql/delegate/show_tables.go:
```diff
@@ -43,53 +43,37 @@ func (d *delegator) delegateShowTables(n *tree.ShowTables) (tree.Statement, erro
+     , (
+        SELECT estimated_row_count
+          FROM crdb_internal.table_row_statistics
+         WHERE table_id = pc.oid::int
```
Does this handle when the requested table ID doesn't exist in the stats table?

ekalinin replied:
Could you elaborate, please.
Currently, If there's no requested table ID in the stats table then we return NULL.

rohany reviewed (changes requested):
I'm wondering if background stats collection could result in these test's row counts flaking. Could that happen @rytaft?

Also, what are the thoughts on displaying this row count only if requested, like `SHOW TABLES ... WITH ROW COUNT` or something @awoods187? 

Also, this row count is irrelevant for virtual tables, so I would prefer if we reported null rather than 0 for them.

rytaft reviewed (commented):
> I'm wondering if background stats collection could result in these test's row counts flaking.

Yes, it could. I think we currently have automatic stats collection disabled for logic tests, but I think the hope was that we'd be able to re-enable it eventually. You should probably explicitly disable automatic stats collection with `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false` in any file where an automatic stats run could cause flakiness.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52203#-:-MDywyma8btWw46rsmHI:b48wagg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @ekalinin, @knz, and @rohany)



<!-- Sent from Reviewable.io -->

ekalinin commented:
> Also, what are the thoughts on displaying this row count only if requested, like `SHOW TABLES ... WITH ROW COUNT` or something?
> 

Yeah, good point. Will update `parser` & `ShowTables` struct.

> Also, this row count is irrelevant for virtual tables, so I would prefer if we reported null rather than 0 for them.

Done.

rohany commented:
> Yeah, good point. Will update parser & ShowTables struct.

Hold off on updating until we decide whether we should do this or not -- it's a good amount of work to be going back and forth.

ekalinin commented:
> > Yeah, good point. Will update parser & ShowTables struct.
> 
> Hold off on updating until we decide whether we should do this or not -- it's a good amount of work to be going back and forth.

Ok :)

ekalinin commented:
Rebased. Fixed some tests.

rohany commented on pkg/sql/crdb_internal.go:
```diff
@@ -340,6 +341,62 @@ CREATE TABLE crdb_internal.tables (
+				tableID := tree.DInt(table.ID)
+				rowCount := tree.DNull
+				// For Virtual Tables report NULL row count.
+				if !table.IsVirtualTable() {
```
Do virtual tables show up in the stats? I think we also want to just default row count to NULL if we cannot find the table in statMap

ekalinin replied:
Yep, virtual tables will show up in `crdb_internal.table_row_statistics` (even if there's no any rows for it in `system.table_statistics`).
All virtual tables will have NULL number of rows as you asked here:

- https://github.com/cockroachdb/cockroach/pull/52203#pullrequestreview-461284517

> Also, this row count is irrelevant for virtual tables, so I would prefer if we reported null rather than 0 for them.

rohany reviewed (changes requested):
I would also improve the release note here to say the new column name, and how it is calculated (for the docs team to document).

I'd also like you to add a test where you select from this new table and ensure that only tables the current user has permission to see show up

ekalinin commented:
> I would also improve the release note here to say the new column name, and how it is calculated (for the docs team to document).

Done.

> 
> I'd also like you to add a test where you select from this new table and ensure that only tables the current user has permission to see show up

Done:

- https://github.com/cockroachdb/cockroach/pull/52203/files#diff-cdf15f008dd1eaef8d7900a2736e8a97R530

Test user can see only virtual tables (`estimated_row_count` is NULL) and tables user has access to (`estimated_row_count` >= 0).

ekalinin commented:
This PR still has one failed test:

```
------- Stdout: -------
=== RUN   TestTenantLogic/3node-tenant/table
=== PAUSE TestTenantLogic/3node-tenant/table
=== CONT  TestTenantLogic/3node-tenant/table
        --- FAIL: TestTenantLogic/3node-tenant/table (9.41s)
            logic.go:2308: 
                
                ../../sql/logictest/testdata/logic_test/table:652: 
                expected success, but found
                (XX000) internal error: unexpected concurrency for a flow that was forced to be planned locally
                distsql_running.go:415: in Run()
                DETAIL: stack trace:
                github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:415: Run()
                github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_stats.go:272: planAndRunCreateStats()
                github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:502: func2()
                github.com/cockroachdb/cockroach/pkg/kv/db.go:707: func1()
                github.com/cockroachdb/cockroach/pkg/kv/txn.go:811: exec()
                github.com/cockroachdb/cockroach/pkg/kv/db.go:706: Txn()
                github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:490: Resume()
                github.com/cockroachdb/cockroach/pkg/jobs/registry.go:851: stepThroughStateMachine()
                github.com/cockroachdb/cockroach/pkg/jobs/registry.go:987: func1()
                github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347: func1()
                runtime/asm_amd64.s:1357: goexit()
                
                NOTE: internal errors may have more details in logs. Use -show-logs.
            logic.go:2644: 
                ../../sql/logictest/testdata/logic_test/table:652: error while processing
            logic.go:2644: pq: internal error: unexpected concurrency for a flow that was forced to be planned locally
```

It failed here:

- https://github.com/cockroachdb/cockroach/pull/52203/files#diff-cdf15f008dd1eaef8d7900a2736e8a97R652-R654

And looks like it linked to this issue:

- https://github.com/cockroachdb/cockroach/issues/50049

rohany commented:
Oh, it's started to fail because you've added a stats collection run. Just add a `# LogicTest: !3node-tenant(50049)` to the top of the `table` logictest file.

ekalinin commented:
> Oh, it's started to fail because you've added a stats collection run. Just add a `# LogicTest: !3node-tenant(50049)` to the top of the `table` logictest file.

Thanks! Added.

rohany commented:
This looks good to me! Thanks for your contribution.

bors r=rohany

Amruta-Ranade commented:
@ekalinin Thank you for contributing to CockroachDB this year. As a token of our appreciation, we would like to send you a gift. Please DM me on our [community Slack](https://cockroa.ch/slack) @Amruta so I can send you a link. (If you don’t want a gift, we also have a charitable contribution choice.) All orders need to be in by December 13, so please contact me as soon as possible :)
